lib/microcbor-derive/src/lib.rs

```diff
@@ -0,0 +1,1030 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+extern crate proc_macro;
+use proc_macro::TokenStream;
+use quote::{ToTokens, format_ident, quote};
+use syn::{
+    Attribute, DataEnum, DataStruct, DeriveInput, Generics, Ident, LitStr,
+    Visibility, parse_macro_input,
+};
+
+/// Derives an implementation of the [`Encode`] and `StaticCborLen` traits for the
+/// annotated `struct` or `enum` type.
+///
+/// All fields of the deriving type must implement the [`Encode`] and
+/// `StaticCborLen` traits, with the following exceptions:
+///
+/// - If the field is annotated with the `#[cbor(skip)]` attribute,
+///   it need not implement any traits, as it is skipped.
+/// - If the field is annotated with the `#[cbor(flatten)]` attribute,
+///   it must instead implement the [`EncodeFields`] trait.
+///
+/// Because fields must implement `StaticCborLen`, the maximum length in bytes
+/// of the encoded representation can be computed at compile-time.
+///
+/// # Encoding
+///
+/// The generated CBOR is encoded as follows:
+///
+/// - Structs with named fields, and struct-like enum variants, are encoded
+///   as CBOR maps of strings to values. The keys in the encoded map are the
+///   string representations of the Rust identifier names of the encoded
+///   fields, unless overridden by the `#[cbor(rename = "..")]` attribute.
+/// - Structs with unnamed fields ("tuple structs") and enum variants with
+///   unnamed fields are encoded as CBOR arrays of the values of those
+///   fields, in declaration order.
+/// - If a tuple struct or tuple-like enum variant has only a single field,
+///   it is encoded "transparently", i.e. as the CBOR value of that field,
+///   rather than as a single-element array.
+/// - Unit enum variants are encoded as strings. By default, the string
+///   representation is the Rust identifier name of the variant, unless
+///   overridden by the `#[cbor(rename = "...")]` attribute.
+///
+///   Someday, I may add a way to encode enum variants as their `repr`
+///   values, but I haven't done that yet.
+///
+/// ## Enum Variant ID Encoding
+///
+/// The `#[cbor(variant_id = "...")]` attribute may be placed on an enum type to
+/// encode its variants with a variant ID field, similar to [`serde`'s
+/// "internally tagged" enum representations][serde-tagged].
+///
+/// **Note**: We use the (somewhat unfortunate) terminology "variant ID" rather
+/// than "tag", because in CBOR, the term "tag" refers to a [completely
+/// different thing][cbor-tags].
+///
+/// If the `#[cbor(variant_id = "id_field_name")]` attribute is present, any
+/// variant of the enum will additionally encode a key-value pair where the
+/// key is the provided variant ID field name, and the value is the variant's
+/// name (or the value of a `#[cbor(rename = "...")]` attribute on that
+/// variant if one is present).
+///
```

Comment:
Explain what happens if it's not present?  I guess fields within the variant are just plopped into the CBOR object with no tag?